File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.py
n, k = map(int,input().split())
arr = [[0,0]]
for _ in range(n):
  arr.append(list(map(int,input().split())))
d = [[0 for _ in range(k+1)] for _ in range(n+1)]
for i in range(1, n+1):
  for j in range(1, k+1):
    weight = arr[i][0]
    value = arr[i][1]
    if j < weight:
      d[i][j] = d[i-1][j]
    else:
      d[i][j] = max(d[i-1][j] ,d[i-1][j-weight]+value)
print(d[n][k])

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.cc
#include<iostream>
#include <algorithm>
#include <vector>
using namespace std;

int n, k;
int d[101][100001];
vector<pair<int, int>> bag;

int main() {
	cin >> n >> k;
	bag.push_back({ 0,0 });
	for (int i = 0; i < n; i++) {
		int w, v;
		cin >> w >> v;
		bag.push_back({ w,v });
	}

	for (int i = 1; i <= n; i++) {
		for (int j = 1; j <= k; j++) {
			int weight = bag[i].first;
			int value = bag[i].second;

			if (j < weight) {
				d[i][j] = d[i - 1][j];
			}
			else {
				d[i][j] = max(d[i - 1][j - weight] + value, d[i - 1][j]);
			}
		}
	}
	cout << d[n][k];
	return 0;
}